### PR TITLE
Auto imports

### PR DESCRIPTION
**Pull request by williamluke4**
@borislit This requires quite a bit of work but let me know your thoughts

### CODE REVIEW

File: src/modules/jsx.ts [hunk 1]
```diff
@@ -86,38 +93,45 @@ export function wrapWithComponent(fullPath, jsx): ProcessedSelection {
...
-            componentProperties.memberProps.add(path.parent.property.name);
+          if (astPath.node.property.name === 'props') {
+            componentProperties.memberProps.add(astPath.parent.property.name);
           } else {
-            path.node.property.name = 'props';
-            componentProperties.state.add(path.parent.property.name);
+            astPath.node.property.name = 'props';
+            componentProperties.state.add(astPath.parent.property.name);
           }
+        } else if (t.isThisExpression(astPath.node.object)) {
```
Comment (borislit) [inline src/modules/jsx.ts]: @williamluke4 can you explain what case does this part handle?

File: package.json [hunk 2]
```diff
@@ -96,13 +96,15 @@
...
         "@babel/types": "^7.1.2",
+        "@types/prettier": "^1.15.2",
         "gitignore-globs": "^0.1.1",
         "gitignore-to-glob": "^0.3.0",
         "glob": "^7.1.3",
         "line-column": "^1.0.2",
         "mkdirp": "^0.5.1",
         "prepend": "^1.0.2",
         "prepend-file": "^1.3.1",
+        "prettier": "^1.15.3",
```
Comment (borislit) [inline package.json]: Not handling the formatting is by design. Formatting is very opinionated and every code base has its own. I think we should lets IDEs handle that

**Comment by borislit** [comment]
@williamluke4 can you describe overall solution? I think that the version version should be minimal and just handle the React import. Then, once that works, lets expand it to other imports. The more I think of it, the more I understand importing React has plenty of complexity in an of it self

**Comment by williamluke4** [comment]
@borislit Might be best to explain a bit more over a call.

File: src/modules/jsx.ts (hunk 1 above)
Comment (williamluke4) [reply src/modules/jsx.ts]: glean

File: package.json (hunk 2 above)
Comment (williamluke4) [reply package.json]: How is it handled by the IDE? And I'm pretty sure prettier can pick up the workspaces config. E.g from tslint, eslint ect

File: src/get-imports.ts [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+  const visitor = {
+    JSXIdentifier(path) {
+      if (path.container.type === 'JSXOpeningElement') {
+        imports = [...imports, path.node.name];
+      }
+    },
+  };
+
+  // include React from Source
+  imports = [...imports, 'React'];
```
Comment (williamluke4) [inline src/get-imports.ts]: @borislit React is added to the imports on the exported components imports here as it (react) wouldn't be picked up in the selected codes JSXOpeningElements

Comment (williamluke4) [reply src/get-imports.ts]: Just for clarification, this function search the selected code for JSXOpeningElements. Then searches that files' imports for the found JSXElements

File: src/modules/jsx.ts (hunk 1 above)
Comment (williamluke4) [reply src/modules/jsx.ts]: @borislit Let me know if that cleared it up

File: package.json (hunk 2 above)
Comment (williamluke4) [reply package.json]: How about
```glean.prettier: true``` with ```default:false```

Comment (borislit) [reply package.json]: the problem is that this still doesnt solve the issue of venturing into the formatting territory. I dont see much point in that. Lets folks run their local formatter of their choice. We dont need to do it for the IDE

File: src/modules/jsx.ts (hunk 1 above)
Comment (borislit) [reply src/modules/jsx.ts]: Oh Amazing!!! Its bug fix.  Can you create a separate PR? I want to keep sem-ver and release it ASAP. Thanks! @williamluke4

**Comment by borislit** [comment]
@williamluke4 hey! whats up? You've been missed. Any plans to keep working on this? I've added a basic React import here: https://github.com/wix/vscode-glean/blob/useState-support/src/modules/jsx.ts#L46
WDYT?
Also, i'd LOVE a separate PR for the bugfix we discussed. I can create one with ur code if you'd like

**Pull request closed without merging**